bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/handler/GardenaAccountHandler.java

```diff
@@ -51,80 +60,168 @@
+
     /**
      * Initializes the GardenaSmart account.
+     * This method is called on a thread.
      */
-    private void initializeGardena() {
-        final GardenaAccountHandler instance = this;
-        scheduler.execute(() -> {
-            try {
-                GardenaConfig gardenaConfig = getThing().getConfiguration().as(GardenaConfig.class);
-                logger.debug("{}", gardenaConfig);
-
-                String id = getThing().getUID().getId();
-                gardenaSmart = new GardenaSmartImpl(id, gardenaConfig, instance, scheduler, httpClientFactory,
-                        webSocketFactory);
-                final GardenaDeviceDiscoveryService discoveryService = this.discoveryService;
-                if (discoveryService != null) {
-                    discoveryService.startScan(null);
-                    discoveryService.waitForScanFinishing();
-                }
-                updateStatus(ThingStatus.ONLINE);
-            } catch (GardenaException ex) {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, ex.getMessage());
-                disposeGardena();
-                if (ex.getStatus() == 429) {
-                    // if there was an error 429 (Too Many Requests), wait for 24 hours before trying again
-                    scheduleReinitialize(REINITIALIZE_DELAY_HOURS_LIMIT_EXCEEDED, TimeUnit.HOURS);
+    private synchronized void initializeGardena() {
+        try {
+            GardenaConfig gardenaConfig = getThing().getConfiguration().as(GardenaConfig.class);
+            logger.debug("{}", gardenaConfig);
+
+            String id = getThing().getUID().getId();
+            gardenaSmart = new GardenaSmartImpl(id, gardenaConfig, this, scheduler, httpClientFactory,
+                    webSocketFactory);
+            final GardenaDeviceDiscoveryService discoveryService = this.discoveryService;
+            if (discoveryService != null) {
+                discoveryService.startScan(null);
+                discoveryService.waitForScanFinishing();
+            }
+            reInitializationCausedBy429 = false;
+            updateStatus(ThingStatus.ONLINE);
+        } catch (GardenaException ex) {
+            logger.warn("{}", ex.getMessage());
+
+            long delaySecs;
+
+            synchronized (reInitializationCodeLock) {
+                Duration delay;
+                boolean isHttp429Error = (HttpStatus.TOO_MANY_REQUESTS_429 == ex.getStatus());
+                if (isHttp429Error) {
+                    delay = REINITIALIZE_DELAY_HOURS_LIMIT_EXCEEDED;
                 } else {
-                    // otherwise reinitialize after 120 seconds
-                    scheduleReinitialize(REINITIALIZE_DELAY_SECONDS, TimeUnit.SECONDS);
+                    Instant now = Instant.now();
+                    Instant notBeforeTime = lastApiCallTime().plus(REINITIALIZE_DELAY_MINUTES_BACK_OFF)
+                            .plus(REINITIALIZE_DELAY_MINUTES_BACK_OFF);
```

Comment:
Why add the delay twice?

Reply:
We need to limit the API call rate to one call every 15 minutes. There are three API calls that must be spread over time, namely 1) the 'lastAPICall', 2) the call that caused this current error, and 3) the yet to-be-scheduled reconnection call. So essentially we have to schedule call 3) to not be less than 30 minutes after call 1)..

Reply:
Whether or not 2) should count into the delay depends on the exception type: If it's a HTTP error (`ex.getStatus()` returns a positive value) that's likely the case, but in case of e.g. a network issue it's likely not the case.

My suggestion would be something like this:
```
if (ex.getStatus() > 0) {
    // HTTP error -> be safe and assume the server counts it against the request rate limit
    lastApiCallTimeUpdate();
}
notBeforeTime = lastApiCallTime().plus(REINITIALIZE_DELAY_MINUTES_BACK_OFF);
```

That way the logic would be easier to understand.

Reply:
Good point. Will do.